### PR TITLE
Avoid here-documents in systemd mount generator

### PR DESCRIPTION
**Pull request by lhuedepohl**
### Motivation and Context
The systemd zfs-mount-generator script fails on my openSUSE systems, as there is not yet a writeable temporary file system available at the time it runs. Thus bash bails out with an error,

  'cannot create temp file for here-document: Read-only file system',

on the here documents in zfs-mount-generator.

### Description
The simple fix is to change these into many individual echo statements.

### How Has This Been Tested?
I use it myself since a few weeks on several machines. The (backported) patch is also active on my project on the [openSUSE Build Service](https://build.opensuse.org/package/show/home:lorenz:filesystems/zfs) where I build zfs packages for openSUSE and which is used by several other people.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by behlendorf** [comment]
@aerusso would you mind commenting on this.

**Comment by InsanePrawn** [comment]
Well here's something I didn't know about - bash uses temp files for heredocs?!

While the echo solution looks okayish to me, my limited sense of shell script aesthetics tells me to write it 'at once' instead of appending multiple echoes to the file repeatedly, so maybe something like this?
```sh
( echo "abc"
...
echo "last line") > "${dest_norm}/${keyloadunit}"
```
saves us a couple syscalls at least.

**Comment by aerusso** [comment]
Wow, I bumped into this in my first version of this, but somehow wasn't able to reproduce the issue. We can avoid the multiple calls with something like

```sh
printf "# Automatically generated by zfs-mount-generator\n"\
"\n"\
"[Unit]\n"\
... > ...
```

**Comment by lhuedepohl** [comment]
I guess one could also do

```
echo "First line
second line
third line" > file
```

**Comment by mschilli87** [comment]
@lhuedepohl:
> I guess one could also do
> 
> ```
> echo "First line
> second line
> third line" > file
> ```

I think especially when planning to support more than just Linux, [`printf` is the safer bet](https://unix.stackexchange.com/q/65803).

---
Also, IMHO this whole change might be worth documenting inline to avoid future more elegant refactors using here-docs. :wink:

**Comment by behlendorf** [comment]
@lhuedepohl I'd suggest we go with either the `printf` or single `echo` command approach you mentioned above.  It agree we should also add a comment that here-docs were intentionally avoided and shouldn't be used.  Please go ahead and refresh this PR when you get a chance.

**Comment by rlaager** [comment]
Portability is only important in the sense that /bin/sh is not always bash. So as long as it works in a POSIX she’ll, we should be good. We may need to rewrite this in C anyway, but fixing this now is still valuable.

**Comment by lhuedepohl** [comment]
Rewrote as requested with single echo and multi-line strings. A ran the `checkbashisms` tool on a little test script with such a construction and it was fine with it - for what it's worth.

File: etc/systemd/system-generators/zfs-mount-generator.in [hunk 1]
```diff
@@ -104,8 +104,13 @@ process_line() {
       fi
 
       # Generate the key-load .service unit
-      cat > "${dest_norm}/${keyloadunit}" << EOF
-# Automatically generated by zfs-mount-generator
+      #
+      # Note: It is tempting to use a `<<EOF` style here-document for this, but
+      #   bash requires a writable /tmp or $TMPDIR for that. This is not always
+      #   available early during boot
```
Comment (rlaager) [inline etc/systemd/system-generators/zfs-mount-generator.in]: nit: add a period at the end of this sentence

Comment (lhuedepohl) [reply etc/systemd/system-generators/zfs-mount-generator.in]: Here you go :)

**Pull request merged**